community/neo4j/src/test/java/recovery/RecoveryCleanupIT.java

```diff
@@ -0,0 +1,194 @@
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package recovery;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.neo4j.graphdb.DependencyResolver;
+import org.neo4j.graphdb.GraphDatabaseService;
+import org.neo4j.graphdb.Label;
+import org.neo4j.graphdb.Transaction;
+import org.neo4j.kernel.NeoStoreDataSource;
+import org.neo4j.kernel.api.labelscan.LabelScanStore;
+import org.neo4j.kernel.impl.transaction.log.checkpoint.CheckPointer;
+import org.neo4j.kernel.impl.transaction.log.checkpoint.SimpleTriggerInfo;
+import org.neo4j.kernel.internal.DatabaseHealth;
+import org.neo4j.kernel.internal.GraphDatabaseAPI;
+import org.neo4j.kernel.monitoring.Monitors;
+import org.neo4j.test.Barrier;
+import org.neo4j.test.Race;
+import org.neo4j.test.TestGraphDatabaseFactory;
+import org.neo4j.test.rule.TestDirectory;
+
+import static org.junit.Assert.fail;
+
+public class RecoveryCleanupIT
+{
+    @Rule
+    public TestDirectory testDirectory = TestDirectory.testDirectory();
+
+    private TestGraphDatabaseFactory factory = new TestGraphDatabaseFactory();
+    private Barrier.Control recoveryCompleteBarrier;
+    private GraphDatabaseService db;
+    private File storeDir;
+    private ExecutorService executor = Executors.newFixedThreadPool( 2 );
+    private Label label = Label.label( "label" );
+
+    @Before
+    public void setup()
+    {
+        storeDir = testDirectory.graphDbDir();
+
+        Monitors monitors = new Monitors();
+        recoveryCompleteBarrier = new Barrier.Control();
+        LabelScanStore.Monitor recoveryBarrierMonitor = new RecoveryBarrierMonitor( recoveryCompleteBarrier );
+        monitors.addMonitorListener( recoveryBarrierMonitor );
+        factory.setMonitors( monitors );
+    }
+
+    @Test
+    public void recoveryCleanupShouldBlockCheckpoint() throws Throwable
+    {
+        // GIVEN
+        AtomicReference<Throwable> error = new AtomicReference<>();
```

Comment:
```
Future<?> checkpointFuture = executor.submit( (Callable<Void>) () -> {
                checkpoint( db );
                return null;
            } );
```
for example will allow you to get future with checkpoint inside :) but this is also fine it can be like it is now